lib/truncate.js

```diff
@@ -10,192 +10,222 @@ exports.error = truncError
+                  max = opts.truncateKeywordsAt
+                  break
+              }
+              break
+          }
+          break
+
+        case 'process':
+          if (path[1] === 'title') {
             max = opts.truncateKeywordsAt
-            break
-        }
-        break
+          }
+          break
+
+        case 'system':
+          switch (path[1]) {
+            case 'architecture':
+            case 'hostname':
+            case 'platform':
+              max = opts.truncateKeywordsAt
+              break
+          }
+          break
+      }
+
+      return truncate(value, max)
     }
+  })
+}
 
-    return truncate(value, max)
-  }, true)
+function onObject (value, key, path, isNew) {
+  if (isHeader(path)) return value.toString()
+  return isNew ? {} : '[Circular]'
 }
 
 function truncTransaction (trans, opts) {
-  trans.name = truncate(String(trans.name), opts.truncateKeywordsAt)
-  trans.type = truncate(String(trans.type), opts.truncateKeywordsAt)
-  trans.result = truncate(String(trans.result), opts.truncateKeywordsAt)
+  const result = breadthFilter(trans, {
+    onObject,
+    onValue (value, key, path) {
+      if (typeof value !== 'string') {
+        if (isHeader(path)) return String(value)
 
-  breadthFilter(trans, (value, key, path) => {
-    if (typeof value !== 'string') {
-      return value
-    }
-
-    let max = opts.truncateStringsAt
-    switch (path[0]) {
-      case 'name':
-      case 'type':
-      case 'result':
         return value
+      }
 
-      case 'id':
-      case 'trace_id':
-      case 'parent_id':
-        max = opts.truncateKeywordsAt
-        break
+      let max = opts.truncateStringsAt
+      switch (path[0]) {
+        case 'name':
+        case 'type':
+        case 'result':
+          value = String(value)
+          max = opts.truncateKeywordsAt
+          break
+
+        case 'id':
+        case 'trace_id':
+        case 'parent_id':
+          max = opts.truncateKeywordsAt
+          break
 
-      case 'context':
-        max = contextLength(path, opts)
-        break
+        case 'context':
+          max = contextLength(path, opts)
+          break
+      }
+
+      return truncate(value, max)
     }
+  })
 
-    return truncate(value, max)
-  }, true)
+  return Object.assign({
+    name: 'undefined',
+    type: 'undefined',
+    result: 'undefined'
+  }, result)
 }
 
 function truncSpan (span, opts) {
-  span.name = truncate(String(span.name), opts.truncateKeywordsAt)
-  span.type = truncate(String(span.type), opts.truncateKeywordsAt)
+  const result = breadthFilter(span, {
+    onObject,
+    onValue (value, key, path) {
+      if (typeof value !== 'string') {
+        return value
+      }
 
-  breadthFilter(span, (value, key, path) => {
-    if (typeof value !== 'string') {
-      return value
-    }
+      let max = opts.truncateStringsAt
+      switch (path[0]) {
+        case 'name':
+        case 'type':
+          value = String(value)
```

Comment:
No need to stringify the value. It will happen inside `truncate()` as well. So these 3 cases can just be combined with the following group of cases.

Reply:
fixed